sql: Index routine create virtual tables by routine IDs.

OriSavir opened the pull request:
Previously, the internal routine virtual tables,
`crdb_internal.create_function_statements` and
`crdb_internal.create_procedure_statements` were
not indexed. As such, in the functions to query for a routine create statement by the routine ID,
which are necessary for `SHOW CREATE ALL ROUTINES`, the query was on a virtual table that was not indexed and, thus, inefficient. This commit addresses that by adding a function that can be used to build rows of the virtual tables directly from the routine IDs and, thus, can act as a populate function for virtual indexes. This function was added onto the virtual tables and now the query is more efficient.

Epic: CRDB-49582
Release note: None

OriSavir committed:
sql: Index routine create virtual tables by routine IDs.

Previously, the internal routine virtual tables,
`crdb_internal.create_function_statements` and
`crdb_internal.create_procedure_statements` were
not indexed. As such, in the functions to query for
a routine create statement by the routine ID,
which are necessary for `SHOW CREATE ALL ROUTINES`,
the query was on a virtual table that was not indexed
and, thus, inefficient. This commit addresses that
by adding a function that can be used to build rows
of the virtual tables directly from the routine IDs
and, thus, can act as a populate function for virtual indexes.
This function was added onto the virtual tables and
now the query is more efficient.

Epic: CRDB-49582
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148217)
<!-- Reviewable:end -->

rafiss reviewed (commented):
nice work! one thing we should do in this PR is test to see if it had a performance effect. the way we can do that is by adding a testcase in the `rttanalysis` package. 

take a look at pkg/bench/rttanalysis/virtual_table_bench_test.go and add a test case like this.

```
		// This test checks the performance of the crdb_internal table used by
		// SHOW CREATE ALL ROUTINES.
		{
			Name: "show_create_all_routines",
			Setup: buildNFunctions(100), // Implement this function; see buildNTables for a similar one.
			Stmt: `SHOW CREATE ALL ROUTINES`,
		},
```

then you can run this test using `./dev bench pkg/bench/rttanalysis -f=BenchmarkVirtualTableQueries/show_create_all_routines --verbose`

pro-tip: if you add this in a commit before the commit that adds this index, you can run the test before and after your change so you can easily see if it had an effect. (remember you can use interactive rebase to reorder the commits.)

OriSavir committed:
rttanalysis: benchmark `SHOW CREATE ALL ROUTINES`

Before, there was no benchmark for performanceo of
`SHOW CREATE ALL ROUTINES`. Now, a benchmark was
added within `virtual_table_bench_test`.
This test builds functions and executes the new
SQL statement to benchmark the time it takes
to query for all the functions.
This sort of benchmark will be helpful
for determining whether indexing changes
improved performance.

Epic: CRDB-49582
Release note: None

rafiss reviewed (commented):
nice job with the benchmark and improvement! this is close to being ready

(i won't mark it with "request changes" so that while i'm gone, you can get an approval from someone else.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-:-OSf9tdT3Cge8nP1WAA5:b-65j9g9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/bench/rttanalysis/virtual_table_bench_test.go` line 18 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-OSf9ZRS0h6TDlxz4Nkm:-OSf9ZRS0h6TDlxz4Nkn:bedw981) ([raw file](https://github.com/cockroachdb/cockroach/blob/c488f92d88e96237d0338033c861eca9fd8e06ec/pkg/bench/rttanalysis/virtual_table_bench_test.go#L18)):*
> ```Go
> 			Name: "select crdb_internal.tables with 1 fk",
> 			Setup: `
> CREATE TABLE t1 (i INT PRI	MARY KEY);
> ```

nit: this whitespace change is breaking the test.
___
*[`pkg/sql/crdb_internal.go` line 3927 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-OSf9i5J9vXm8sS0RH3f:-OSf9i5J9vXm8sS0RH3g:b47g6k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c488f92d88e96237d0338033c861eca9fd8e06ec/pkg/sql/crdb_internal.go#L3927)):*
> ```Go
> 	// optimize the queries for the create statements output in `SHOW CREATE ALL ROUTINES`.
> 	fnID := descpb.ID(tree.MustBeDInt(unwrappedConstraint))
> 	fnDescs, err := p.Descriptors().ByIDWithoutLeased(p.txn).WithoutNonPublic().Get().Descs(ctx, []descpb.ID{fnID})
> ```

nit: instead of using `.Descs()` use `.Function()`. that will return it with the correct type
___
*[`pkg/sql/crdb_internal.go` line 3938 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-OSf9fQ39vD6VZOw0MzV:-OSf9fQ39vD6VZOw0MzW:b-ln20ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/c488f92d88e96237d0338033c861eca9fd8e06ec/pkg/sql/crdb_internal.go#L3938)):*
> ```Go
> 	sc, err := p.Descriptors().ByIDWithoutLeased(p.txn).WithoutNonPublic().Get().Schema(ctx, scID)
> 	if err != nil || sc == nil {
> 		if errors.Is(err, catalog.ErrDescriptorNotFound) || sc == nil {
> ```

nit:  we don't want this not found handling when fetching the schema -- we should only have the check for DescriptorNotFound when looking up by function ID (as you already have above)


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
we will also need to update the file pkg/bench/rttanalysis/testdata/benchmark_expectations

you can do that with: `./dev test --rewrite pkg/bench/rttanalysis -f='TestBenchmarkExpectation/VirtualTableQueries/show_create' --test-args '-rewrite-iterations=10'`

also, would you be able to add a benchmark for show_create_all_triggers as well? (feel free to handle that in a separate PR if you prefer)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-:-OSfGYme7SlR16vBNy-X:b6awhtx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @OriSavir)



<!-- Sent from Reviewable.io -->

OriSavir reviewed (commented):
Done! Saw an ~10x improvement.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-:-OSeatCF2ZP_Re_USrzZ:b4y1ht3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
> Reviewed 3 of 3 files at r3, 3 of 3 files at r5, 1 of 3 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148217)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @OriSavir)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 134 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-OStDyMT9t0lp9vUwm1u:-OStDyMT9t0lp9vUwm1v:b2fqp87) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ca525952f42e6af92cbc748fa675bf28757906/pkg/bench/rttanalysis/testdata/benchmark_expectations#L134)):*
> ```
> 1,VirtualTableQueries/select_crdb_internal.invalid_objects_with_1_fk
> 1,VirtualTableQueries/select_crdb_internal.tables_with_1_fk
> 0,VirtualTableQueries/show_create_all_routines
> ```

This seems too amazing? I would expect us to at least look up the database descriptor


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-:-OStTM1G9WIOg3mKfo2-:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @OriSavir)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 134 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/148217#-OStDyMT9t0lp9vUwm1u:-OStTEfV360dCV3g81OI:b-307o5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/61ca525952f42e6af92cbc748fa675bf28757906/pkg/bench/rttanalysis/testdata/benchmark_expectations#L134)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This seems too amazing? I would expect us to at least look up the database descriptor
</blockquote></details>

We confirmed in a debugger the looks ups are happening, so somehow everything is cached


<!-- Sent from Reviewable.io -->

OriSavir commented:
bors r+
